### PR TITLE
clippy-driver: pass all args to rustc if --rustc is present

### PR DESCRIPTION
**Pull request by matthiaskrgr**
changelog: clippy-driver: pass all args to rustc if --rustc is present

### CODE REVIEW

File: src/driver.rs [hunk 1]
```diff
@@ -299,6 +299,21 @@ pub fn main() {
         rustc_driver::catch_fatal_errors(move || {
             let mut orig_args: Vec<String> = env::args().collect();
 
+            if orig_args.get(1) == Some(&"rustc".into()) {
+                // first (0) arg is the bin path, second arg is "rustc"
+                // pass all succeeding args to rustc
+                let args_for_rustc = &orig_args[2..].to_vec();
+
+                let exitstatus = Command::new("rustc")
```
Comment (jsgf) [inline src/driver.rs]: No, this doesn't help. `rustc` isn't necessarily in the path, and even if it were, doesn't necessarily have anything to do with the version of rustc `clippy-driver` was built against. If I wanted to know the version of the `rustc` executable, I could just run `rustc --version`.

**Comment by flip1995** [comment]
What we could test for starters, is if the first argument is `rustc`, that we immediately return `rustc_driver::run_compiler(&args, DefaultCallbacks, None, None)`, where `args` is `orig_args` without `clippy_driver`. This should implement `clippy_driver rustc ...` = `rustc ...`-linked-with-clippy. If that works, we can carefully go through the `args` modifications in this closure and determine what should be kept when `rustc` is passed.

**Comment by flip1995** [comment]
I just realized, that the `clippy-driver rustc` idea doesn't work:
https://github.com/rust-lang/rust-clippy/blob/06f0ab03ef63869eb159cf7d57d5d5e826ff10ca/src/driver.rs#L353-L355

`clippy-driver rustc` is run, when running `cargo clippy`...

**Comment by flip1995** [comment]
What do you think about `clippy-driver --rustc`?

**Comment by flip1995** [comment]
This would be the patch for adding this https://github.com/rust-lang/rust-clippy/pull/5178#issuecomment-587058988:
```diff
diff --git a/src/driver.rs b/src/driver.rs
index 097b796e..b280169d 100644
--- a/src/driver.rs
+++ b/src/driver.rs
@@ -299,6 +299,12 @@ pub fn main() {
         rustc_driver::catch_fatal_errors(move || {
             let mut orig_args: Vec<String> = env::args().collect();
 
+            if let Some(pos) = orig_args.iter().position(|arg| arg == "--rustc") {
+                orig_args.remove(pos);
+                orig_args[0] = "rustc".to_string();
+                return rustc_driver::run_compiler(&orig_args, &mut DefaultCallbacks, None, None);
+            }
+
             if orig_args.iter().any(|a| a == "--version" || a == "-V") {
                 let version_info = rustc_tools_util::get_version_info!();
                 println!("{}", version_info);
```

**Comment by matthiaskrgr** [comment]
>I just realized, that the clippy-driver rustc idea doesn't work

Urgh that sucks...
Imo the intuition with `--arguments` is that the order does not really matter, so `clippy-driver --rustc --version` should behave like `clippy-driver --version --rustc` which is not the case here :(
But I don't have any better ideas right now..

**Comment by flip1995** [comment]
The patch I posted in https://github.com/rust-lang/rust-clippy/pull/5178#issuecomment-588335005 takes care of this and **everything** will be passed to `rustc`, when `--rustc` is passed. The only thing, that didn't work as expected is `clippy-driver --rustc some_file.rs`, which should be `rustc some_file.rs`, but threw `error: counldn't find crate for std`.

**Comment by flip1995** [comment]
Friendly ping @matthiaskrgr. Do you think we can/should move forward with this?

**Comment by matthiaskrgr** [comment]
looks like the `error: counldn't find crate for std.` problem is caused by missing `--sysroot` arg when calling `rustc`, this is actually injected a little bit later in the code: https://github.com/rust-lang/rust-clippy/blob/master/src/driver.rs#L386

**Comment by flip1995** [comment]
Can you add a test script, that compares output of some `clippy-driver --rustc [some_args]` commands with `rustc [some_args]`? Similar to the `.github/driver.sh` script. (Or you can just add tests for this there)

**Comment by matthiaskrgr** [comment]
Added test for  `clippy-driver --rustc --version --verbose` vs `rustc --version --verbose` to driver.sh

**Comment by flip1995** [comment]
This test would also be nice to have:

```bash
echo "fn main() {}" > tmp.rs
clippy-driver --rustc tmp.rs
```

and maybe a test where `--rustc` ist not the first argument.

**Comment by matthiaskrgr** [comment]
Hmm, `clippy-driver --version --rustc` prints `rustc --version`, this is a bug, right?

File: .github/driver.sh [hunk 2]
```diff
@@ -31,4 +31,9 @@ diff normalized.stderr tests/ui/cstring.stderr
 SYSROOT=`rustc --print sysroot`
 diff <(LD_LIBRARY_PATH=${SYSROOT}/lib ./target/debug/clippy-driver --rustc --version --verbose) <(rustc --version --verbose)
 
+# we can't run 2 rustcs on the same file at the same time
+CLIPPY=`LD_LIBRARY_PATH=${SYSROOT}/lib ./target/debug/clippy-driver tests/driver/main.rs`
```
Comment (flip1995) [inline .github/driver.sh:35]:
```suggestion
CLIPPY=`LD_LIBRARY_PATH=${SYSROOT}/lib ./target/debug/clippy-driver --rustc tests/driver/main.rs`
```

I think you forgot the `--rustc` arg here.

Comment (matthiaskrgr) [reply .github/driver.sh:35]: Oh, right.

**Comment by matthiaskrgr** [comment]
CI was failing because I was testing against a file that I forgot to git-commit... :laughing:

File: .github/driver.sh [hunk 3]
```diff
@@ -26,4 +26,14 @@ unset CARGO_MANIFEST_DIR
 sed -e "s,tests/ui,\$DIR," -e "/= help/d" cstring.stderr > normalized.stderr
 diff normalized.stderr tests/ui/cstring.stderr
 
+
+# make sure "clippy-driver --rustc --arg" and "rustc --arg" behave the same
+SYSROOT=`rustc --print sysroot`
+diff <(LD_LIBRARY_PATH=${SYSROOT}/lib ./target/debug/clippy-driver --rustc --version --verbose) <(rustc --version --verbose)
+
+# we can't run 2 rustcs on the same file at the same time
```
[the remark this answers is not in the document]
Comment (matthiaskrgr) [reply .github/driver.sh:36]: just in case I put the file into `target` so it does not show up in "git status"

**Pull request merged**